File: src/php/core/src/Renderer.php
<?php

declare(strict_types=1);

namespace DiceBear;

use DiceBear\Prng\Fnv1a;
use DiceBear\Style\Canvas;
use DiceBear\Style\Element;
use DiceBear\Utils\Initials;
use DiceBear\Utils\License;
use DiceBear\Utils\Number;
use DiceBear\Utils\Xml;

/**
 * Walks a style's element tree and turns it into the final SVG markup.
 *
 * The renderer is single-use: it accumulates `<defs>` entries and per-render
 * caches across method calls, so a fresh instance is required per avatar.
 */
class Renderer
{
    private Style $style;
    private Resolver $resolver;
    /** @var array<string, string> */
    private array $defs = [];
    private ?string $cachedSeedHash = null;
    private ?string $cachedInitials = null;

    public function __construct(Style $style, Resolver $resolver)
    {
        $this->style = $style;
        $this->resolver = $resolver;
    }

    /**
     * Builds the complete SVG document for the avatar.
     */
    public function render(): string
    {
        $canvas = $this->style->canvas();
        $background = $this->renderBackground($canvas);
        $body = $this->renderElements($canvas->elements());

        // Order matters: scale and flip around center, then rotate, translate,
        // and finally clip with border radius (outermost wrapper).
        $body = $this->applyScale($body, $canvas);
        $body = $this->applyFlip($body, $canvas);
        $body = $this->applyRotate($body, $canvas);
        $body = $this->applyTranslate($body, $canvas);
        $body = $this->applyBorderRadius($background . $body, $canvas);

        $metadata = License::xml($this->style->meta());
        $defs = count($this->defs) > 0
            ? '<defs>' . implode('', array_values($this->defs)) . '</defs>'
            : '';
        $size = $this->resolver->size();

        $title = $this->resolver->title();
        $escapedTitle = $title !== null ? Xml::escape($title) : null;

        $width = Number::format($canvas->width());
        $height = Number::format($canvas->height());
        $attrs = [
            'xmlns="http://www.w3.org/2000/svg"',
            "viewBox=\"0 0 {$width} {$height}\"",
        ];

        $rootAttributes = $this->renderAttributes($this->style->attributes());

        if ($rootAttributes !== '') {
            $attrs[] = ltrim($rootAttributes);
        }

        if ($escapedTitle !== null) {
            $attrs[] = 'role="img"';
            $attrs[] = "aria-label=\"{$escapedTitle}\"";
        } else {
            $attrs[] = 'aria-hidden="true"';
        }

        if ($size !== null) {
            $sizeValue = Number::format($size);
            $attrs[] = "width=\"{$sizeValue}\"";
            $attrs[] = "height=\"{$sizeValue}\"";
        }

        $titleElement = $escapedTitle !== null ? "<title>{$escapedTitle}</title>" : '';

        $svg = '<svg ' . implode(' ', $attrs) . '>' . $metadata . $defs . $titleElement . $body . '</svg>';

        if ($this->resolver->idRandomization()) {
            $svg = $this->randomizeIds($svg);
        }

        return $svg;
    }

    /**
     * Wraps `$body` in a flip transform when `flip` is set to anything other
     * than `'none'`.
     */
    private function applyFlip(string $body, Canvas $canvas): string
    {
        $flip = $this->resolver->flip();

        if ($flip === 'none') {
            return $body;
        }

        $w = Number::format($canvas->width());
        $h = Number::format($canvas->height());

        // @phpstan-ignore match.unhandled
        $transform = match ($flip) {
            'horizontal' => "translate({$w}, 0) scale(-1, 1)",
            'vertical' => "translate(0, {$h}) scale(1, -1)",
            'both' => "translate({$w}, {$h}) scale(-1, -1)",
        };

        return "<g transform=\"{$transform}\">{$body}</g>";
    }

    /**
     * Wraps `$body` in a uniform scale transform around the canvas center
     * when the option differs from `1`.
     */
    private function applyScale(string $body, Canvas $canvas): string
    {
        $scale = $this->resolver->scale();

        if ($scale === 1.0) {
            return $body;
        }

        $cx = $canvas->width() / 2;
        $cy = $canvas->height() / 2;
        $cxValue = Number::format($cx);
        $cyValue = Number::format($cy);
        $scaleValue = Number::format($scale);
        $negCx = Number::format(-$cx);
        $negCy = Number::format(-$cy);

        return "<g transform=\"translate({$cxValue}, {$cyValue}) scale({$scaleValue}) translate({$negCx}, {$negCy})\">{$body}</g>";
    }

    /**
     * Clips `$body` to the canvas rectangle (rounded when `borderRadius` is
     * non-zero) and registers the corresponding `clipPath` in `<defs>`. The
     * clip is always applied so transformed content cannot bleed past the
     * canvas bounds, regardless of the consumer's `overflow` setting.
     */
    private function applyBorderRadius(string $body, Canvas $canvas): string
    {
        $radius = $this->resolver->borderRadius();
        $id = 'clip-' . $this->hashSeed();

        $rx = Number::format(($radius / 100) * $canvas->width());
        $ry = Number::format(($radius / 100) * $canvas->height());
        $width = Number::format($canvas->width());
        $height = Number::format($canvas->height());

        $this->defs[$id] = "<clipPath id=\"{$id}\"><rect width=\"{$width}\" height=\"{$height}\" rx=\"{$rx}\" ry=\"{$ry}\"/></clipPath>";

        return "<g clip-path=\"url(#{$id})\">{$body}</g>";
    }

    /**
     * Wraps `$body` in a rotation around the canvas center when `rotate` is
     * non-zero.
     */
    private function applyRotate(string $body, Canvas $canvas): string
    {
        $rotate = $this->resolver->rotate();

        if ($rotate === 0.0) {
            return $body;
        }

        $cx = Number::format($canvas->width() / 2);
        $cy = Number::format($canvas->height() / 2);
        $rotateValue = Number::format($rotate);

        return "<g transform=\"rotate({$rotateValue}, {$cx}, {$cy})\">{$body}</g>";
    }

    /**
     * Wraps `$body` in a translate transform when either `translateX` or
     * `translateY` is non-zero. Offsets are interpreted as percentages of
     * the canvas dimensions.
     */
    private function applyTranslate(string $body, Canvas $canvas): string
    {
        $translateX = $this->resolver->translateX();
        $translateY = $this->resolver->translateY();

        if ($translateX === 0.0 && $translateY === 0.0) {
            return $body;
        }

        $x = Number::format(($translateX / 100) * $canvas->width());
        $y = Number::format(($translateY / 100) * $canvas->height());

        return "<g transform=\"translate({$x}, {$y})\">{$body}</g>";
    }

    /**
     * Returns a `<rect>` filling the canvas with the resolved background
     * color, or an empty string when no background colors are configured.
     */
    private function renderBackground(Canvas $canvas): string
    {
        $colors = $this->resolver->color('background');

        if (count($colors) === 0) {
            return '';
        }

        $fill = Xml::escape($this->resolveColorReference('background'));
        $width = Number::format($canvas->width());
        $height = Number::format($canvas->height());

        return "<rect width=\"{$width}\" height=\"{$height}\" fill=\"{$fill}\"/>";
    }

    /**
     * Suffixes every `id` declaration and reference with a random hex string
     * so that multiple instances of the same avatar do not collide in a
     * shared document. Uses `random_int()` intentionally — a PRNG-derived
     * suffix would produce the same ID for the same seed.
     */
    private function randomizeIds(string $svg): string
    {
        $suffix = str_pad(dechex(random_int(0, 0xffffff)), 6, '0', STR_PAD_LEFT);
        $ids = [];

        if (preg_match_all('/\bid="([^"]+)"/', $svg, $matches)) {
            $ids = array_unique($matches[1]);
        }

        if (count($ids) === 0) {
            return $svg;
        }

        $escaped = array_map(fn($id) => preg_quote($id, '/'), $ids);
        $pattern = '/(id="|url\(#|href="#)(' . implode('|', $escaped) . ')("|\))/';

        return preg_replace_callback($pattern, function ($match) use ($suffix) {
            return $match[1] . $match[2] . '-' . $suffix . $match[3];
        }, $svg);
    }

    /**
     * Renders a list of elements and concatenates their markup.
     *
     * @param list<Element> $elements
     */
    private function renderElements(array $elements): string
    {
        return implode('', array_map(fn($el) => $this->renderElement($el), $elements));
    }

    /**
     * Dispatches a single element to the renderer for its type.
     */
    private function renderElement(Element $element): string
    {
        // @phpstan-ignore match.unhandled
        return match ($element->type()) {
            'element' => $this->renderSvgElement($element),
            'text' => $this->renderTextElement($element),
            'component' => $this->renderComponentElement($element),
        };
    }

    /**
     * Renders an SVG element. The special `defs` name diverts children into
     * the shared `<defs>` block.
     *
     * Element names and attribute names are not escaped here — they are
     * validated by StyleValidator against a strict allowlist schema (no
     * `<script>`, no event handlers). Values are escaped via `Xml::escape()`.
     */
    private function renderSvgElement(Element $element): string
    {
        $name = $element->name();

        if ($name === null) {
            return '';
        }

        if ($name === 'defs') {
            foreach ($element->children() as $child) {
                $rendered = $this->renderElement($child);

                if (strlen($rendered) > 0) {
                    $attrs = $child->attributes();
                    $id = isset($attrs['id']) && is_string($attrs['id']) ? $attrs['id'] : '_' . count($this->defs);
                    $this->defs[$id] = $rendered;
                }
            }

            return '';
        }

        $attrs = $this->renderAttributes($element->attributes());
        $children = $this->renderElements($element->children());

        if (strlen($children) === 0) {
            return "<{$name}{$attrs}/>";
        }

        return "<{$name}{$attrs}>{$children}</{$name}>";
    }

    /**
     * Renders a text element by escaping its resolved value.
     */
    private function renderTextElement(Element $element): string
    {
        $value = $element->value();

        return $value !== null ? Xml::escape($this->resolveValue($value)) : '';
    }

    /**
     * Resolves a component reference to a chosen variant and emits a `<use>`
     * pointing at a `<defs>` entry that holds the variant body. Aliases of the
     * same source component sharing a variant — and identical components
     * referenced more than once — therefore produce a single `<defs>` entry
     * referenced by every `<use>`, never duplicated SVG markup.
     *
     * Any `attributes` on the component reference are written to the emitted
     * `<use>` tag. A user-supplied `transform` is prepended to the
     * per-component transforms so it acts as the outer (placement) transform,
     * with the style's translate/rotate/scale applied inside it.
     */
    private function renderComponentElement(Element $element): string
    {
        $componentName = $element->name();

        if (!is_string($componentName)) {
            return '';
        }

        $variantName = $this->resolver->variant($componentName);

        if ($variantName === null) {
            return '';
        }

        $components = $this->style->components();
        $component = $components[$componentName] ?? null;

        if ($component === null) {
            return '';
        }

        $variants = $component->variants();
        $variant = $variants[$variantName] ?? null;

        if ($variant === null) {
            return '';
        }

        $id = "{$component->sourceName()}-{$variantName}-{$this->hashSeed()}";

        if (!array_key_exists($id, $this->defs)) {
            $body = $this->renderElements($variant->elements());

            $this->defs[$id] = "<g id=\"{$id}\">{$body}</g>";
        }

        $transforms = $this->buildTransforms($component);
        $userAttributes = $element->attributes();
        $mergedAttributes = $userAttributes;

        if (count($transforms) > 0) {
            $userTransform = $userAttributes['transform'] ?? null;
            $allParts = is_string($userTransform) && $userTransform !== ''
                ? [$userTransform, ...$transforms]
                : $transforms;

            $mergedAttributes = $userAttributes ?? [];
            $mergedAttributes['transform'] = implode(' ', $allParts);
        }

        $attrs = $this->renderAttributes($mergedAttributes);

        return "<use{$attrs} href=\"#{$id}\"/>";
    }

    /**
     * Returns the per-component SVG `transform` fragments derived from the
     * component's translate, rotate, and scale options. Translate values are
     * percentages of the component canvas dimensions, matching the
     * semantics of the user-facing `translateX` / `translateY` options.
     *
     * The fragments are ordered so that, when joined into a single
     * `transform` attribute, the scale is the rightmost (innermost)
     * transform — applied first to a point, followed by rotate, then
     * translate.
     *
     * @return list<string>
     */
    private function buildTransforms(Style\Component $component): array
    {
        ['rotate' => $rotate, 'translateX' => $translateX, 'translateY' => $translateY, 'scale' => $scale] =
            $this->resolver->componentTransform($component->name());

        if ($translateX === 0.0 && $translateY === 0.0 && $rotate === 0.0 && $scale === 1.0) {
            return [];
        }

        $transforms = [];
        $cx = $component->width() / 2;
        $cy = $component->height() / 2;
        $cxValue = Number::format($cx);
        $cyValue = Number::format($cy);

        if ($translateX !== 0.0 || $translateY !== 0.0) {
            $x = Number::format(($translateX / 100) * $component->width());
            $y = Number::format(($translateY / 100) * $component->height());
            $transforms[] = "translate({$x}, {$y})";
        }

        if ($rotate !== 0.0) {
            $rotateValue = Number::format($rotate);
            $transforms[] = "rotate({$rotateValue}, {$cxValue}, {$cyValue})";
        }

        if ($scale !== 1.0) {
            $scaleValue = Number::format($scale);
            $negCx = Number::format(-$cx);
            $negCy = Number::format(-$cy);
            $transforms[] = "translate({$cxValue}, {$cyValue}) scale({$scaleValue}) translate({$negCx}, {$negCy})";
        }

        return $transforms;
    }

    /**
     * Serializes an attribute map to a leading space-prefixed string suitable
     * for inlining into a tag. Returns an empty string when there are no
     * attributes to render.
     *
     * @param array<string, mixed>|null $attributes
     */
    private function renderAttributes(?array $attributes): string
    {
        if ($attributes === null) {
            return '';
        }

        $parts = [];

        foreach ($attributes as $key => $value) {
            if ($value === null) {
                continue;
            }

            $parts[] = "{$key}=\"" . Xml::escape($this->resolveAttributeValue($value)) . '"';
        }

        if (count($parts) === 0) {
            return '';
        }

        return ' ' . implode(' ', $parts);
    }

    /**
     * Resolves a single attribute value: literal strings pass through, color
     * and variable references are dereferenced through the option resolver.
     *
     * @param string|array<string, mixed> $value
     */
    private function resolveAttributeValue(string|array $value): string
    {
        if (is_string($value)) {
            return $value;
        }

        if (($value['type'] ?? null) === 'color') {
            return $this->resolveColorReference($value['name']);
        }

        return $this->resolveVariable($value['name']);
    }

    /**
     * Resolves a named color into either a hex string (solid fill / single
     * color) or a `url(#…)` gradient reference, registering the gradient in
     * `<defs>` as a side effect.
     */
    private function resolveColorReference(string $name): string
    {
        $colors = $this->resolver->color($name);
        $fill = $this->resolver->colorFill($name);

        if ($fill === 'solid' || count($colors) <= 1) {
            return $colors[0] ?? 'none';
        }

        return $this->buildGradientDef($name, $colors, $fill);
    }

    /**
     * Builds the `<linearGradient>` or `<radialGradient>` for the given color
     * definition, registers it in `<defs>`, and returns its `url(#…)`
     * reference.
     *
     * @param list<string> $colors
     */
    private function buildGradientDef(string $name, array $colors, string $fill): string
    {
        $rotation = $this->resolver->colorAngle($name);
        $id = "{$name}-color-{$this->hashSeed()}";
        $tag = $fill === 'linear' ? 'linearGradient' : 'radialGradient';
        $rotationValue = Number::format($rotation);
        $rotateAttr = $rotation !== 0.0
            ? " gradientTransform=\"rotate({$rotationValue}, 0.5, 0.5)\""
            : '';

        $stops = [];
        $count = count($colors);

        foreach ($colors as $i => $color) {
            $offset = Number::format($i / ($count - 1) * 100);
            $stops[] = "<stop offset=\"{$offset}%\" stop-color=\"" . Xml::escape($color) . '"/>';
        }

        $this->defs[$id] = "<{$tag} id=\"{$id}\"{$rotateAttr}>" . implode('', $stops) . "</{$tag}>";

        return "url(#{$id})";
    }

    /**
     * Resolves an element value to its final string form. Literal strings
     * pass through; variable references are dereferenced.
     *
     * @param string|array<string, mixed> $value
     */
    private function resolveValue(string|array $value): string
    {
        if (is_string($value)) {
            return $value;
        }

        if (($value['type'] ?? null) === 'variable') {
            return $this->resolveVariable($value['name']);
        }

        return '';
    }

    /**
     * Resolves a built-in variable reference to its current value.
     */
    private function resolveVariable(string $name): string
    {
        // @phpstan-ignore match.unhandled
        return match ($name) {
            'initial' => mb_substr($this->initials(), 0, 1),
            'initials' => $this->initials(),
            'fontWeight' => Number::format($this->resolver->fontWeight()),
            'fontFamily' => $this->resolver->fontFamily(),
        };
    }

    /**
     * Returns the seed-derived initials, cached after the first call.
     */
    private function initials(): string
    {
        return $this->cachedInitials ??= Initials::fromSeed($this->resolver->seed());
    }

    /**
     * Returns the FNV-1a hex hash of the seed, cached after the first call.
     * The value is used to derive stable but unique IDs for `<defs>` entries.
     */
    private function hashSeed(): string
    {
        return $this->cachedSeedHash ??= Fnv1a::hex($this->resolver->seed());
    }
}

File: src/python/core/src/dicebear/renderer.py
"""Walks a style's element tree and turns it into the final SVG markup."""

from __future__ import annotations

import random
import re
from typing import Any

from .prng import Fnv1a
from .resolver import Resolver
from .style import Style
from .style_def.canvas import Canvas
from .style_def.component import Component
from .style_def.element import Element
from .utils.initials import Initials
from .utils.license import License
from .utils.number import Number
from .utils.xml import Xml

_ID_PATTERN = re.compile(r'\bid="([^"]+)"')


class Renderer:
    """Walks a style's element tree and turns it into the final SVG markup.

    The renderer is single-use: it accumulates ``<defs>`` entries and per-render
    caches across method calls, so a fresh instance is required per avatar.
    """

    def __init__(self, style: Style, resolver: Resolver) -> None:
        self._style = style
        self._resolver = resolver
        self._defs: dict[str, str] = {}
        self._cached_seed_hash: str | None = None
        self._cached_initials: str | None = None

    def render(self) -> str:
        """Build the complete SVG document for the avatar."""
        canvas = self._style.canvas()
        background = self._render_background(canvas)
        body = self._render_elements(canvas.elements())

        # Order matters: scale and flip around center, then rotate, translate,
        # and finally clip with border radius (outermost wrapper).
        body = self._apply_scale(body, canvas)
        body = self._apply_flip(body, canvas)
        body = self._apply_rotate(body, canvas)
        body = self._apply_translate(body, canvas)
        body = self._apply_border_radius(background + body, canvas)

        metadata = License.xml(self._style.meta())
        defs = "<defs>" + "".join(self._defs.values()) + "</defs>" if self._defs else ""
        size = self._resolver.size()

        title = self._resolver.title()
        escaped_title = Xml.escape(title) if title is not None else None

        width = Number.format(canvas.width())
        height = Number.format(canvas.height())
        attrs = [
            'xmlns="http://www.w3.org/2000/svg"',
            f'viewBox="0 0 {width} {height}"',
        ]

        root_attributes = self._render_attributes(self._style.attributes())

        if root_attributes != "":
            attrs.append(root_attributes.lstrip())

        if escaped_title is not None:
            attrs.append('role="img"')
            attrs.append(f'aria-label="{escaped_title}"')
        else:
            attrs.append('aria-hidden="true"')

        if size is not None:
            attrs.append(f'width="{Number.format(size)}"')
            attrs.append(f'height="{Number.format(size)}"')

        title_element = (
            f"<title>{escaped_title}</title>" if escaped_title is not None else ""
        )

        svg = (
            "<svg "
            + " ".join(attrs)
            + ">"
            + metadata
            + defs
            + title_element
            + body
            + "</svg>"
        )

        if self._resolver.id_randomization():
            svg = self._randomize_ids(svg)

        return svg

    def _apply_flip(self, body: str, canvas: Canvas) -> str:
        """Wrap ``body`` in a flip transform when ``flip`` is not ``'none'``."""
        flip = self._resolver.flip()

        if flip == "none":
            return body

        w = Number.format(canvas.width())
        h = Number.format(canvas.height())

        transform = {
            "horizontal": f"translate({w}, 0) scale(-1, 1)",
            "vertical": f"translate(0, {h}) scale(1, -1)",
            "both": f"translate({w}, {h}) scale(-1, -1)",
        }[flip]

        return f'<g transform="{transform}">{body}</g>'

    def _apply_scale(self, body: str, canvas: Canvas) -> str:
        """Wrap ``body`` in a uniform scale around center when scale != 1."""
        scale = self._resolver.scale()

        if scale == 1:
            return body

        cx = canvas.width() / 2
        cy = canvas.height() / 2
        cx_value = Number.format(cx)
        cy_value = Number.format(cy)
        scale_value = Number.format(scale)
        neg_cx = Number.format(-cx)
        neg_cy = Number.format(-cy)

        return (
            f'<g transform="translate({cx_value}, {cy_value}) '
            f"scale({scale_value}) "
            f'translate({neg_cx}, {neg_cy})">{body}</g>'
        )

    def _apply_border_radius(self, body: str, canvas: Canvas) -> str:
        """Clip ``body`` to the (rounded) canvas rectangle via a ``clipPath``.

        The clip is always applied so transformed content cannot bleed past the
        canvas bounds, regardless of the consumer's overflow setting.
        """
        radius = self._resolver.border_radius()
        id_ = "clip-" + self._hash_seed()

        rx = Number.format((radius / 100) * canvas.width())
        ry = Number.format((radius / 100) * canvas.height())
        width = Number.format(canvas.width())
        height = Number.format(canvas.height())

        self._defs[id_] = (
            f'<clipPath id="{id_}"><rect width="{width}" height="{height}" '
            f'rx="{rx}" ry="{ry}"/></clipPath>'
        )

        return f'<g clip-path="url(#{id_})">{body}</g>'

    def _apply_rotate(self, body: str, canvas: Canvas) -> str:
        """Wrap ``body`` in a rotation around center when ``rotate`` is non-zero."""
        rotate = self._resolver.rotate()

        if rotate == 0:
            return body

        cx = Number.format(canvas.width() / 2)
        cy = Number.format(canvas.height() / 2)
        rotate_value = Number.format(rotate)

        return f'<g transform="rotate({rotate_value}, {cx}, {cy})">{body}</g>'

    def _apply_translate(self, body: str, canvas: Canvas) -> str:
        """Wrap ``body`` in a translate when translateX/translateY is non-zero.

        Offsets are interpreted as percentages of the canvas dimensions.
        """
        tx = self._resolver.translate_x()
        ty = self._resolver.translate_y()

        if tx == 0 and ty == 0:
            return body

        x = Number.format((tx / 100) * canvas.width())
        y = Number.format((ty / 100) * canvas.height())

        return f'<g transform="translate({x}, {y})">{body}</g>'

    def _render_background(self, canvas: Canvas) -> str:
        """Return a ``<rect>`` with the resolved background color, or ``''``."""
        colors = self._resolver.color("background")

        if len(colors) == 0:
            return ""

        fill = Xml.escape(self._resolve_color_reference("background"))
        width = Number.format(canvas.width())
        height = Number.format(canvas.height())

        return f'<rect width="{width}" height="{height}" fill="{fill}"/>'

    def _randomize_ids(self, svg: str) -> str:
        """Suffix every ``id`` declaration and reference with a random hex string.

        Uses ``random`` intentionally — a PRNG-derived suffix would produce the
        same ID for the same seed.
        """
        suffix = format(random.randint(0, 0xFFFFFF), "06x")

        ids = list(dict.fromkeys(_ID_PATTERN.findall(svg)))

        if len(ids) == 0:
            return svg

        escaped = [re.escape(i) for i in ids]
        pattern = re.compile(r'(id="|url\(#|href="#)(' + "|".join(escaped) + r')("|\))')

        return pattern.sub(
            lambda m: m.group(1) + m.group(2) + "-" + suffix + m.group(3), svg
        )

    def _render_elements(self, elements: list[Element]) -> str:
        """Render a list of elements and concatenate their markup."""
        return "".join(self._render_element(el) for el in elements)

    def _render_element(self, element: Element) -> str:
        """Dispatch a single element to the renderer for its type."""
        element_type = element.type()

        if element_type == "element":
            return self._render_svg_element(element)

        if element_type == "text":
            return self._render_text_element(element)

        if element_type == "component":
            return self._render_component_element(element)

        return ""

    def _render_svg_element(self, element: Element) -> str:
        """Render an SVG element; the special ``defs`` name diverts children.

        Element and attribute names are not escaped here — they are validated by
        StyleValidator against a strict allowlist schema. Values are escaped via
        :meth:`Xml.escape`.
        """
        name = element.name()

        if name is None:
            return ""

        if name == "defs":
            for child in element.children():
                rendered = self._render_element(child)

                if len(rendered) > 0:
                    attrs = child.attributes()
                    child_id = attrs.get("id") if attrs is not None else None
                    key = (
                        child_id
                        if isinstance(child_id, str)
                        else "_" + str(len(self._defs))
                    )
                    self._defs[key] = rendered

            return ""

        attrs_str = self._render_attributes(element.attributes())
        children = self._render_elements(element.children())

        if len(children) == 0:
            return f"<{name}{attrs_str}/>"

        return f"<{name}{attrs_str}>{children}</{name}>"

    def _render_text_element(self, element: Element) -> str:
        """Render a text element by escaping its resolved value."""
        value = element.value()

        return Xml.escape(self._resolve_value(value)) if value is not None else ""

    def _render_component_element(self, element: Element) -> str:
        """Resolve a component reference to a variant and emit a ``<use>``.

        Aliases of the same source component sharing a variant — and identical
        components referenced more than once — produce a single ``<defs>`` entry
        referenced by every ``<use>``. A user-supplied ``transform`` is prepended
        to the per-component transforms.
        """
        component_name = element.name()

        if not isinstance(component_name, str):
            return ""

        variant_name = self._resolver.variant(component_name)

        if variant_name is None:
            return ""

        component = self._style.components().get(component_name)

        if component is None:
            return ""

        variant = component.variants().get(variant_name)

        if variant is None:
            return ""

        id_ = f"{component.source_name()}-{variant_name}-{self._hash_seed()}"

        if id_ not in self._defs:
            body = self._render_elements(variant.elements())
            self._defs[id_] = f'<g id="{id_}">{body}</g>'

        transforms = self._build_transforms(component)
        user_attributes = element.attributes()
        merged_attributes: dict[str, Any] | None = user_attributes

        if len(transforms) > 0:
            user_transform = (
                user_attributes.get("transform") if user_attributes else None
            )
            if isinstance(user_transform, str) and user_transform != "":
                all_parts = [user_transform, *transforms]
            else:
                all_parts = transforms

            merged_attributes = {
                **(user_attributes or {}),
                "transform": " ".join(all_parts),
            }

        attrs_str = self._render_attributes(merged_attributes)

        return f'<use{attrs_str} href="#{id_}"/>'

    def _build_transforms(self, component: Component) -> list[str]:
        """Return the per-component SVG ``transform`` fragments.

        Ordered so that, joined into one ``transform`` attribute, the scale is
        the rightmost (innermost) transform — applied first, then rotate, then
        translate.
        """
        t = self._resolver.component_transform(component.name())
        rotate = t["rotate"]
        translate_x = t["translateX"]
        translate_y = t["translateY"]
        scale = t["scale"]

        if translate_x == 0 and translate_y == 0 and rotate == 0 and scale == 1:
            return []

        transforms: list[str] = []
        cx = component.width() / 2
        cy = component.height() / 2
        cx_value = Number.format(cx)
        cy_value = Number.format(cy)

        if translate_x != 0 or translate_y != 0:
            x = Number.format((translate_x / 100) * component.width())
            y = Number.format((translate_y / 100) * component.height())
            transforms.append(f"translate({x}, {y})")

        if rotate != 0:
            rotate_value = Number.format(rotate)
            transforms.append(f"rotate({rotate_value}, {cx_value}, {cy_value})")

        if scale != 1:
            scale_value = Number.format(scale)
            neg_cx = Number.format(-cx)
            neg_cy = Number.format(-cy)
            transforms.append(
                f"translate({cx_value}, {cy_value}) scale({scale_value}) "
                f"translate({neg_cx}, {neg_cy})"
            )

        return transforms

    def _render_attributes(self, attributes: dict[str, Any] | None) -> str:
        """Serialize an attribute map to a leading space-prefixed string."""
        if attributes is None:
            return ""

        parts: list[str] = []

        for key, value in attributes.items():
            if value is None:
                continue

            parts.append(
                f'{key}="' + Xml.escape(self._resolve_attribute_value(value)) + '"'
            )

        if len(parts) == 0:
            return ""

        return " " + " ".join(parts)

    def _resolve_attribute_value(self, value: str | dict[str, Any]) -> str:
        """Resolve a single attribute value: literals pass through; references
        are dereferenced through the option resolver.
        """
        if isinstance(value, str):
            return value

        if value.get("type") == "color":
            return self._resolve_color_reference(value["name"])

        return self._resolve_variable(value["name"])

    def _resolve_color_reference(self, name: str) -> str:
        """Resolve a named color into a hex string or a ``url(#…)`` gradient."""
        colors = self._resolver.color(name)
        fill = self._resolver.color_fill(name)

        if fill == "solid" or len(colors) <= 1:
            return colors[0] if len(colors) > 0 else "none"

        return self._build_gradient_def(name, colors, fill)

    def _build_gradient_def(self, name: str, colors: list[str], fill: str) -> str:
        """Build the linear/radial gradient, register it, return its reference."""
        rotation = self._resolver.color_angle(name)
        id_ = f"{name}-color-{self._hash_seed()}"
        tag = "linearGradient" if fill == "linear" else "radialGradient"
        rotate_attr = (
            f' gradientTransform="rotate({Number.format(rotation)}, 0.5, 0.5)"'
            if rotation != 0
            else ""
        )

        stops: list[str] = []
        count = len(colors)

        for i, color in enumerate(colors):
            offset = Number.format(i / (count - 1) * 100)
            stops.append(
                f'<stop offset="{offset}%" stop-color="' + Xml.escape(color) + '"/>'
            )

        self._defs[id_] = (
            f'<{tag} id="{id_}"{rotate_attr}>' + "".join(stops) + f"</{tag}>"
        )

        return f"url(#{id_})"

    def _resolve_value(self, value: str | dict[str, Any]) -> str:
        """Resolve an element value: literals pass through; variables resolved."""
        if isinstance(value, str):
            return value

        if value.get("type") == "variable":
            return self._resolve_variable(value["name"])

        return ""

    def _resolve_variable(self, name: str) -> str:
        """Resolve a built-in variable reference to its current value."""
        if name == "initial":
            return self._initials()[0:1]

        if name == "initials":
            return self._initials()

        if name == "fontWeight":
            return Number.format(self._resolver.font_weight())

        if name == "fontFamily":
            return self._resolver.font_family()

        return ""

    def _initials(self) -> str:
        """Return the seed-derived initials, cached after the first call."""
        if self._cached_initials is None:
            self._cached_initials = Initials.from_seed(self._resolver.seed())

        return self._cached_initials

    def _hash_seed(self) -> str:
        """Return the FNV-1a hex hash of the seed, cached after the first call."""
        if self._cached_seed_hash is None:
            self._cached_seed_hash = Fnv1a.hex(self._resolver.seed())

        return self._cached_seed_hash
